src/Plutip/Types.purs

```diff
@@ -197,27 +216,138 @@ instance DecodeAeson StopClusterResponse where
+  decodeWallets
+    :: PrivatePaymentKey
+    -> distr
+    -> Array PrivateKeyResponse
+    -> Contract () wallets
 
 instance UtxoDistribution Unit Unit where
   encodeDistribution _ = []
-  decodeWallets _ = Just unit
+  decodeWallets _ _ _ = pure unit
 
 instance UtxoDistribution InitialUTxO KeyWallet where
   encodeDistribution amounts = [ amounts ]
-  decodeWallets [ (PrivateKeyResponse key) ] =
+  decodeWallets _ _ [ PrivateKeyResponse key ] =
     pure $ privateKeysToKeyWallet (PrivatePaymentKey key) Nothing
-  decodeWallets _ = Nothing
+  decodeWallets _ _ _ = wrongNumberPrivateKeysError
 
 instance
   UtxoDistribution restSpec restWallets =>
   UtxoDistribution (InitialUTxO /\ restSpec) (KeyWallet /\ restWallets) where
-  encodeDistribution (amounts /\ rest) =
-    encodeDistribution amounts <> encodeDistribution rest
-  decodeWallets = Array.uncons >>> case _ of
-    Nothing -> Nothing
-    Just { head: PrivateKeyResponse key, tail } ->
-      Tuple (privateKeysToKeyWallet (PrivatePaymentKey key) Nothing) <$>
-        decodeWallets tail
+  encodeDistribution = encodeDistributionRecurse
+  decodeWallets ourKey (distr /\ rest) = decodeWalletsRecurse ourKey distr rest
+
+instance UtxoDistribution InitialUTxOWithStakeKey KeyWallet where
+  encodeDistribution (InitialUTxOWithStakeKey _ amounts) = [ amounts ]
+  decodeWallets ourKey (InitialUTxOWithStakeKey stake _) [ key ] =
+    decodeWallet ourKey key (Just stake)
+  decodeWallets _ _ _ = wrongNumberPrivateKeysError
+
+instance
+  UtxoDistribution restSpec restWallets =>
+  UtxoDistribution (InitialUTxOWithStakeKey /\ restSpec)
+    (KeyWallet /\ restWallets) where
+  encodeDistribution = encodeDistributionRecurse
+  decodeWallets ourKey (distr /\ rest) =
+    decodeWalletsRecurse ourKey distr rest
+
+encodeDistributionRecurse
+  :: forall distr wallets restSpec restWallets
+   . UtxoDistribution distr wallets
+  => UtxoDistribution restSpec restWallets
+  => (distr /\ restSpec)
+  -> Array (Array UtxoAmount)
+encodeDistributionRecurse (distr /\ rest) =
+  encodeDistribution distr <> encodeDistribution rest
+
+-- | Get a key wallet for the current distribution+PrivateKeyResponse
+-- | pair, then recurse and combine
+decodeWalletsRecurse
+  :: forall distr restSpec restWallets
+   . UtxoDistribution distr KeyWallet
+  => UtxoDistribution restSpec restWallets
+  => PrivatePaymentKey
+  -> distr
+  -> restSpec
+  -> Array PrivateKeyResponse
+  -> Contract () (KeyWallet /\ restWallets)
+decodeWalletsRecurse ourKey distr rest = Array.uncons >>> case _ of
+  Nothing -> wrongNumberPrivateKeysError
+  Just { head, tail } -> do
+    wallet <- decodeWallets ourKey distr [ head ]
+    Tuple wallet <$> decodeWallets ourKey rest tail
+
+-- | Get a wallet for a single private key response, and if a stake
+-- | key is provided for it, transfer the funds allocated by plutip to
+-- | the address combining the payment key and the stake key
+decodeWallet
```

Comment:
I think we should use one contract to fund for all addresses at once. We can add multiple `mustPayToPubKeyAddress` constraints.

This will require more code, but it gives linear performance improvement which is important for tests.

Reply:
Sounds like we should just move the transferring logic out of `decodeWallets` into a separate function we call once after decoding the wallets.

I'm not super familiar with transaction size limits, I wonder if we have to worry about making the single transaction too big